net/host.go

```diff
@@ -155,27 +146,56 @@ func NewHost(cfg *Config) (*host, error) {
 		return nil, fmt.Errorf("failed to format bootnodes: %w", err)
 	}
 
+	if len(bns) > 0 {
+		opts = append(opts, libp2p.EnableAutoRelay(autorelay.WithStaticRelays(bns)))
+	}
+
 	// create libp2p host instance
-	h, err := libp2p.New(opts...)
+	basicHost, err := libp2p.New(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// There is libp2p bug when calling `dual.New` with a cancelled context creating a panic,
+	// so we need the extra guard below:
+	// Panic:  https://github.com/jbenet/goprocess/blob/v0.1.4/impl-mutex.go#L99
+	// Caller: https://github.com/libp2p/go-libp2p-kad-dht/blob/v0.17.0/dht.go#L222
+	if cfg.Ctx.Err() != nil {
+		return nil, err
+	}
+
+	// Note on ModeServer: The dual KAD DHT, by default, puts the LAN interface in server mode and
+	// the WAN interface in ModeClient if it is behind a NAT firewall. In our case, even nodes behind
+	// NAT firewalls should be servers, otherwise remote nodes will not be able to connect and list
+	// their offers.
+	dht, err := dual.New(cfg.Ctx, basicHost,
+		dual.DHTOption(kaddht.BootstrapPeers(bns...)),
+		dual.DHTOption(kaddht.Mode(kaddht.ModeServer)),
+	)
 	if err != nil {
 		return nil, err
 	}
 
+	routedHost := routedhost.Wrap(basicHost, dht)
+
 	ourCtx, cancel := context.WithCancel(cfg.Ctx)
 	hst := &host{
 		ctx:        ourCtx,
 		cancel:     cancel,
-		protocolID: fmt.Sprintf("%s/%s/%d", protocolID, cfg.Environment, cfg.EthChainID),
-		h:          h,
+		protocolID: fmt.Sprintf("%s/%s/%s/%d", protocolID, protocolVersion, cfg.Environment, cfg.EthChainID),
+		h:          routedHost,
 		ds:         ds,
 		bootnodes:  bns,
-		queryBuf:   make([]byte, 1024*5),
```

Comment:
why was this removed?

Reply:
The allocations for swap messages and query messages are all in one place now. The buffers are the same size as the messages creating a lot of code simplifications.